core/src/driver/scheduler/evm.rs

```diff
@@ -32,110 +38,119 @@ impl EvmScheduler {
-                    .wait()
-                {
-                    Ok(()) => info!("successfully submitted solution for batch {}", batch_id),
-                    Err(err) => error!(
-                        "failed to submit solution for batch {}: {:?}",
-                        batch_id, err
-                    ),
+    async fn solve(&mut self, batch_id: u32) -> Result<Option<Solution>> {
+        while let Some(time_limit) = self.solver_time_limit(batch_id).await? {
+            info!(
+                "solving for batch {} with time limit {}s",
+                batch_id,
+                time_limit.as_secs_f64(),
+            );
+            match self.driver.solve_batch(batch_id.into(), time_limit).await {
+                Ok(solution) => {
+                    info!("successfully solved batch {}", batch_id);
+                    return Ok(Some(solution));
+                }
+                Err(DriverError::Retry(err)) => {
+                    error!("driver retryable error for batch {}: {:?}", batch_id, err);
+                }
+                Err(DriverError::Skip(err)) => {
+                    error!("driver error for batch {}: {:?}", batch_id, err);
+                    return Ok(None);
                 }
-            }
-            Err(DriverError::Retry(err)) => {
-                error!("driver retryable error for batch {}: {:?}", batch_id, err);
-            }
-            Err(DriverError::Skip(err)) => {
-                error!("driver error for batch {}: {:?}", batch_id, err);
-                self.last_batch = Some(batch_id);
             }
         }
+        warn!("skipping batch {} because we ran out of time", batch_id);
+        Ok(None)
+    }
 
-        Ok(())
+    async fn submit(&self, batch_id: u32, solution: Solution) {
+        // TODO: handle earliest_solution_submit_time .
+        match self.driver.submit_solution(batch_id.into(), solution).await {
+            Ok(()) => info!("successfully submitted solution for batch {}", batch_id),
+            Err(err) => error!(
+                "failed to submit solution for batch {}: {:?}",
+                batch_id, err
+            ),
+        }
     }
 
-    /// Gets the current solving batch ID.
-    ///
-    /// This is the current batch ID minus 1, as the current batch ID is the ID
-    /// of the batch that is currently accepting orders.
-    fn current_solving_batch(&self) -> Result<u32> {
-        Ok(self.exchange.get_current_auction_index().wait()? - 1)
+    /// Wait for and solve the next batch.
+    async fn step(&mut self) -> Result<u32> {
+        let last_batch = match self.last_batch {
+            None => self.current_solving_batch().await?,
+            Some(batch) => batch,
+        };
+        let new_batch = self.wait_for_batch_to_change(last_batch).await?;
+        // TODO: signal healthiness
+        let solution = self.solve(new_batch).await?;
+        self.last_batch = Some(new_batch);
```

Comment:
I've added a comment in the code:
        // The batch is marked as handled so that the next call to step does not handle it again.
        // In the next call `step` will sleep until the batch id changes.

Reply:
Would it be possible to return batchID from step and pass it back into the step function from above?

I'd prefer that over storing state on the instance as it would make it clear that no other part of the code relies on or mutates it.

Reply:
I changed this. The reason for having it was that the code before this PR would retry solving from `start` if `step` returned an error before finishing solving but not during solution submission. Trying to keep this behavior while removing the field is awkward because we would have to return the batch id in both the ok and error case in which case we wouldn't be able to use `?` in `step`.

However, I now realize that the original behavior before splitting stablex_driver to have a solve and submit method was to not retry in this case. This is also how the system scheduler works. So I changed the behavior back to that which allows us to make this change here.